### PR TITLE
cudnn: init cudnn_8_3 at 8.3.0

### PR DESCRIPTION
**Pull request by samuela**
###### Motivation for this change
Followup to https://github.com/NixOS/nixpkgs/pull/158186. This is a necessary step towards addressing the bug https://github.com/google/jax/discussions/9455. This, In conjunction with a bump to the cudnn82 variant of the jaxlib wheel, fixes that issue.

Version 8.3.2 is the latest at the time of writing but the directory structure for the download changed after 8.3.0 so I'll leave updating to 8.3.1 or 8.3.2 to a future PR...

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by samuela** [comment]
friendly ping on this cc @danieldk 

this is currently blocking resolving https://github.com/google/jax/discussions/9455 in nixpkgs's version of jaxlib

**Comment by danieldk** [comment]
> friendly ping on this cc @danieldk
> 
> this is currently blocking resolving [google/jax#9455](https://github.com/google/jax/discussions/9455) in nixpkgs's version of jaxlib

Thanks for the ping, but I currently do not have the time to look at and try PRs.

**Comment by samuela** [comment]
Ok gotcha, who is maintaining/point of contact for the CUDA stuff these days?

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 158218` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>cudnn_cudatoolkit_11 (cudnn_cudatoolkit_11_4)</li>
    <li>cudnn_cudatoolkit_11_0</li>
    <li>katagoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python39Packages.cupy</li>
    <li>python39Packages.tensorflowWithCuda</li>
  </ul>
</details>
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>cudnn_cudatoolkit_11_1</li>
    <li>cudnn_cudatoolkit_11_2</li>
    <li>cudnn_cudatoolkit_11_3</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python39Packages.jaxlibWithCuda</li>
  </ul>
</details>

**Comment by samuela** [comment]
`cudnn_cudatoolkit_11`, `cudnn_cudatoolkit_11_0`, `katagoWithCuda`, `python310Packages.cupy`, and `python39Packages.cupy` all `nix-build` just fine. Not sure why they aren't working in nixpkgs-review (cc @Mic92).

But `python39Packages.tensorflowWithCuda` fails with
```
[18,129 / 22,090] Compiling tensorflow/compiler/xla/service/cpu/runtime_matmul.cc; 91s local ... (192 actions, 191 running)
ERROR: /build/source/tensorflow/compiler/tf2xla/BUILD:1111:11: Linking of rule '//tensorflow/compiler/tf2xla:side_effect_util' failed (Exit 1): ar failed: error executing command /nix/store/bp3qyx8ff4j0g8fzl9q7vvcgww22cr27-gcc-wrapper-9.3.0-merged/bin/ar @bazel-out/k8-opt/bin/tensorflow/compiler/tf2xla/libside_effect_util.pic.a-2.params
INFO: Elapsed time: 338.453s, Critical Path: 138.86s
INFO: 18524 processes: 7719 internal, 10805 local.
FAILED: Build did NOT complete successfully
```
cc maintainers @jyp @abbradar any idea what could be going on here?

**Comment by samuela** [comment]
Ah https://www.tensorflow.org/install/source#gpu indicates that tensorflow 2.7 is tested against cuDNN 8.1 so maybe that's why 8.3 is a problem. But OTOH JAX doesn't work with 8.1... Maybe we'll need two versions?

**Comment by SuperSandro2000** [comment]
Sounds like we need 8.1 and 8.3.

**Comment by samuela** [comment]
Ok, I just force-pushed a new approach in https://github.com/NixOS/nixpkgs/commit/128dee6757dfab78304f42396b727c8fee11ffd1 which keeps all of the existing cuDNN versions and adds 8.3 along with the existing 8.1.

nixpkgs-review was giving me some problems but I've successfully `nix-build`ed each of the following:
* cudnn_7_4_cudatoolkit_10_0
* cudnn_7_6_cudatoolkit_10_1
* cudnn_7_6_cudatoolkit_10_2
* cudnn_8_1_cudatoolkit_11
* cudnn_8_1_cudatoolkit_11_0
* cudnn_8_1_cudatoolkit_11_1
* cudnn_8_1_cudatoolkit_11_3
* cudnn_8_3_cudatoolkit_11_0
* cudnn_8_3_cudatoolkit_11_1
* cudnn_8_3_cudatoolkit_11_2
* cudnn_8_3_cudatoolkit_11_3
* katagoWithCuda
* python310Packages.cupy
* python39Packages.cupy
* python39Packages.tensorflowWithCuda 
* python310Packages.jaxlibWithCuda 
* python39Packages.jaxlibWithCuda

which is everything that this PR touches according to nixpkgs-review.

**Comment by samuela** [comment]
If no one objects, I'll go ahead and merge tomorrow.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -4532,18 +4532,24 @@ with pkgs;
...
-    cudnn_cudatoolkit_10_2
-    cudnn_cudatoolkit_11
-    cudnn_cudatoolkit_11_0
-    cudnn_cudatoolkit_11_1
-    cudnn_cudatoolkit_11_2
-    cudnn_cudatoolkit_11_3
-    cudnn_cudatoolkit_11_4;
-
-  cudnn = cudnn_cudatoolkit_10;
+    cudnn_7_4_cudatoolkit_10_0
```
[the remark this answers is not in the document]
Comment (samuela) [reply pkgs/top-level/all-packages.nix:4535]: pretty unlikely. I can double check and remove them if you'd prefer?

[the remark this answers is not in the document]
Comment (samuela) [reply pkgs/top-level/all-packages.nix]: Done in https://github.com/NixOS/nixpkgs/pull/158218/commits/6052a2f48ce86e8e5808b02b4a197d20a3d73876. There's still lots of things point at old versions that can/should be upgraded, but I'll leave that for a future PR to keep things as small and simple as possible.

**Comment by samuela** [comment]
I was getting some errors like
```
Could not load library libcudnn_ops_infer.so.8. Error: libcublas.so.11: cannot open shared object file: No such file or directory
Please make sure libcudnn_ops_infer.so.8 is in your library path!
Aborted (core dumped)
```
when trying to use `jaxlib-bin` which depends on `cudnn`. Turns out that we need to patch the cudnn RPATH's since some of the cuDNN .so files depend on things in cudatoolkit. I believe this was an oversight in versions past, or that the behavior changed in cuDNN 8.3. In any case, I fixed it in https://github.com/NixOS/nixpkgs/pull/158218/commits/ff869d8bca3ccb3ffb18ce0aab71dfb8a2065ff1.

**Comment by samuela** [comment]
I'll squash and merge tomorrow unless anyone objects.

**Pull request merged**